protocols/relay/src/v2/client/handler.rs

```diff
@@ -0,0 +1,628 @@
+                        renewal_timeout,
+                        pending_msgs,
+                        to_listener,
+                    });
+                }
+            },
+            r => self.reservation = r,
+        }
+
+        // Forward messages to transport listener.
+        if let Some(Reservation::Accepted {
+            pending_msgs,
+            to_listener,
+            ..
+        }) = &mut self.reservation
+        {
+            if !pending_msgs.is_empty() {
+                match to_listener.poll_ready(cx) {
+                    Poll::Ready(Ok(())) => {
+                        if let Err(e) = to_listener
+                            .start_send(pending_msgs.pop_front().expect("Called !is_empty()."))
+                        {
+                            debug!("Failed to sent pending message to listener: {:?}", e);
+                            self.reservation.take();
+                        }
+                    }
+                    Poll::Ready(Err(e)) => {
+                        debug!("Channel to listener failed: {:?}", e);
+                        self.reservation.take();
+                    }
+                    Poll::Pending => {}
+                }
+            }
+        }
+
+        // Deny incoming circuit requests.
+        if let Poll::Ready(Some((src_peer_id, result))) = self.circuit_deny_futs.poll_next_unpin(cx)
+        {
+            match result {
+                Ok(()) => {
+                    return Poll::Ready(ProtocolsHandlerEvent::Custom(
+                        Event::InboundCircuitReqDenied { src_peer_id },
+                    ))
+                }
+                Err(error) => {
+                    return Poll::Ready(ProtocolsHandlerEvent::Custom(
+                        Event::InboundCircuitReqDenyFailed { src_peer_id, error },
+                    ))
+                }
+            }
+        }
+
+        // Send errors to transport.
+        while let Poll::Ready(Some(())) = self.send_error_futs.poll_next_unpin(cx) {}
+
+        // Check status of lend out substreams.
+        loop {
+            match self.alive_lend_out_substreams.poll_next_unpin(cx) {
+                Poll::Ready(Some(Err(oneshot::Canceled))) => {}
+                // TODO: Use void.
+                Poll::Ready(Some(Ok(v))) => void::unreachable(v),
+                Poll::Ready(None) | Poll::Pending => break,
+            }
+        }
+
+        // Update keep-alive handling.
+        if self.reservation.is_none()
+            && self.alive_lend_out_substreams.is_empty()
+            && self.circuit_deny_futs.is_empty()
+        {
+            match self.keep_alive {
+                KeepAlive::Yes => {
+                    self.keep_alive = KeepAlive::Until(Instant::now() + Duration::from_secs(10));
+                }
+                KeepAlive::Until(_) => {}
+                KeepAlive::No => panic!("Handler never sets KeepAlive::No."),
+            }
```

Comment:
We could simplify this by adding:

```rust
impl KeepAlive {
	fn for_atmost(&mut self, duration: Duration) -> Result<(), NotAlive> {
		todo!()	
	}
}
```

```suggestion
            self.keep_alive.for_atmost(Duration::from_secs(10)).expect("Handler never set KeepAlive::No");
```

Reply:
I don't find this more intuitive compared to the current explicit version. That said, in case you feel strongly about it or others prefer the proposed version as well, I am happy to make the changes.

Reply:
> I don't find this more intuitive compared to the current explicit version.

Fair enough :)